### PR TITLE
"Introduction to machine learning with naive bayes" blog post

### PR DESCRIPTION
**Pull request by bobbyshaw**

### CODE REVIEW

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+status: publish
+published: true
+title: Introduction to Machine Learning with Naive Bayes
+categories:
+- development
+---
+
+_Long Read_
+
+2015 [has been said](http://www.bloomberg.com/news/articles/2015-12-08/why-2015-was-a-breakthrough-year-in-artificial-intelligence) to be the year of Artificial Intelligence.  As a keen enthusiast, I am enthralled by all articles surrounding its progression.  AI is a strange term that is often only applied while there is  mysticism around the solution.  However once a problem is solved, it's no longer considered AI (.g. chess computer).  Minor point aside, it's clear to be the source of the next generation of software.
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: "there is  mysticism" -- double space
"(.g. chess computer)" -- e.g.

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+The actual formula can be difficult to get your head around in the first instance so let's go through with our example of language classification.  Can we classify whether a site is French based on in containing the word "mode"?  We know that "mode" is an English word but that it is also French for "fashion". So in order to determine the probability that this is a French site given that it contains the word "mode", we first have to consider the probability that a site would be French anyway, multiplied by the chance that a French site would contain the word "mode" divided by the chance of the word appears in any case.
+
+    P(French|mode) = P(French) x P(mode|French) / P(mode).
+
+In its original form this is:
+
+    P(A|B) = P(A) x P(B|A) / P(B)
+
+Another way that this can be described is that the probability of an class given a feature observed is equal to the ratio of the chance of the class and the chance of the feature being found in that class against the chance that that feature is seen in all classes.
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: "an class"

Also, my head hurts.

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+While naive Bayes is one of the most basic machine learning techniques that does mean there's been plenty of research in how to optimise it and [overcome its assumptions](http://people.csail.mit.edu/jrennie/papers/icml03-nb.pdf).  One of these assumptions is that there are the same number of training examples for each class.  That's not always going to be the case, and so a variation of the model called "Complement Naive Bayes".  This helps to avoid the bias of over-populated classes in the training set by adding a weighting to those that are under-represented.  This is still a simplification over actually getting more training data for those classes.
+
+## PHP Implementation
+
+After some searching for what solutions were out there already, I chose to experiment with Cam Spiers' [Statistical Classifier](https://github.com/camspiers/statistical-classifier) as it seemed the most feature complete.  It provided a Complement Naive Bayes and SVM implementation.
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: What's SVM?

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+```
+
+Once training is complete, a site can be classified by:
+
+```php
+<?php
+$result = $this->classify($site->content);
+```
+
+There were a few other advantages of this library, including it's provision of dependency injection and interfaces to provide your own tokenisation and normalisation (more later).  It also provides the ability to integrate with caching technology, e.g. redis, in order to make access to the trained classifer quicker during execution.
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: "it's"

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+For training I built a laravel based website with commands that would fetch and parse text content from Hivemind's ecommerce dataset.  The training interface loaded the text content parsed and the website in an iframe with two select boxes to confirm.
+
+![Hivemind Classifier Interface](/img/2015/12/hivemind-classifier-interface.jpg)
+
+In terms of presenting sites for training I ran through multiple methodologies.  As time went on and I was evaluating the time/money that went into either myself or third parties performing the training, I wanted to identify ways to create the best training dataset in the shortest of time.  I wagered that the best dataset  had the greatest variation so it would have sufficient sample for each possible result category.  The aim is to have data points that represented the edge cases as well as the sufficient datapoint that enabled trend identification.
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: "dataset  had" -- double space

Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: Check wording of the last sentence too.

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+- **In order of the dataset**
+    - When first creating the system this was a good start while I was focussing on other problems.
+- **Randomise the order**
+    - Dataset is ordered by traffic rank so has a bias towards English and certain categories, e.g. Fashion. Classifying sites in a random order is our next best bet at creating a better sample.
+- **By least accurate**
+    - The laws of probability say that classifying in a random order will, over time, give us a equal exposure to the dataset, but only assuming that the categories in the dataset are of even size.  As the training dataset increase in size I was able to start using it to train a naive bayes implementation and test it's accuracy on the dataset (more on that in the next section).  In doing so, I could then have statistics on the languages and content types that the system was worse at classifying.  By manually classifying sites that the system had classified into groups that we knew that we performed poorly at, it would both provide further training data for poorly classified categories, as well as helping to provide training information for more training data to help correct the mis-classifications.
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: "it's"

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+This brings us onto [cross-validation](https://en.wikipedia.org/wiki/Cross-validation_%28statistics%29). This is where the training dataset is split into `n`-subsets, testing performed on each one and the results averaged in order to get a better estimate of classification accuracy.  It helps to counter-act overfitting and underfitting.
+
+For this problem I chose [2-fold cross-validation](https://en.wikipedia.org/wiki/Cross-validation_%28statistics%29#2-fold_cross-validation) as the testing approach to implement. For this, I continued to split the training dataset into two halves.  The classifier was then trained on the first have and the second half was used to test accuracy.  Then, the classifier was reset and trained on the second half and tested on the first half.  These results could then be averaged to get a better accuracy figure.
+
+# Optimisation
+
+There are two main tasks that our classifier has to undertake, and for those familiar with Big O Notation, these are their complexities:
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: Might be nice to link to https://en.wikipedia.org/wiki/Analysis_of_algorithms here for a definition of Big O.

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+- **Learning/Training (Offline)**
+    - Importing our training data into the system so that it can "learn" and be ready to classifying unseen data.
+    - This process has Big O Notation of `O(n)` where `n` is the largest of either the size of the dataset or the number of different words x number of categories.  In most cases `n` will be the size of the training dataset.  
+- **Classify (Online)**
+    - Processing a document against the trained system and producing a result.
+    - `O(n)` where `n` is the length of the document x number of categories)
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: Trailing bracket

File: _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+This has no little effect on the resource consumption (other than CPU) of the classifier but is done with the aim to improve on accuracy that otherwise would have required further sample data to identify the similarity.
+
+### Tokenisation
+
+For this classification we're using 1-grams, i.e. each token is made up by a single word, you could increase this, e.g. 2 words per taken (2-gram).  Then the additional context serves as more of an identifier. My understanding is that this typically would mean that you would need a larger sample when compared to 1-grams but is typically preferred.
+
+# Alternative Algorithms
+
+I primarily focussed on naive-bayes as a well documented, popular and well regarded alrogithm but there are many more out there that take different approaches which can suit different problems. I experimented also with Support Vector Machine (SVM).
```
Comment (punkstar) [inline _posts/2015-12-21-introduction-to-machine-learning-with-naive-bayes.md]: Ah, that's what SVM means :)

**Comment by bobbyshaw** [comment]
Thanks @punkstar!
